sphinx/environment/__init__.py

```diff
@@ -276,14 +218,17 @@ def __init__(self, app: Sphinx) -> None:
         # objtype index -> (domain, type, objname (localized))
         self._search_index_objnames: dict[int, tuple[str, str, str]] = {}
 
+        # all the registered domains, set by the application
+        self.domains: _DomainsContainer = _DomainsContainer._from_environment(self)
```

Comment:
consider renaming `_DomainsContainer` to just `_Container` (which is namespaced under the `domain` module:

```suggestion
        self.domains: domains._Container = domains._Container._from_environment(self)
```